File: ant-sdk/antd-py/src/antd/_discover.py
"""Daemon port-file discovery for antd.

The antd daemon writes a ``daemon.port`` file on startup containing:
  - Line 1: REST port
  - Line 2: gRPC port
  - Line 3: PID of the daemon process (optional, for staleness detection)

This module reads that file to auto-discover the daemon's listen addresses.
If a PID is present and the process is no longer running, the port file is
considered stale and discovery returns empty results.
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

_PORT_FILE_NAME = "daemon.port"
_DATA_DIR_NAME = "ant"
_SDK_SUBDIR_NAME = "sdk"


def discover_daemon_url() -> str:
    """Return the REST base URL from the daemon port file, or ``""`` on failure."""
    rest, _ = _read_port_file()
    if rest == 0:
        return ""
    return f"http://127.0.0.1:{rest}"


def discover_grpc_target() -> str:
    """Return the gRPC target from the daemon port file, or ``""`` on failure."""
    _, grpc = _read_port_file()
    if grpc == 0:
        return ""
    return f"127.0.0.1:{grpc}"


def _read_port_file() -> tuple[int, int]:
    """Read the daemon.port file and return ``(rest_port, grpc_port)``.

    A single-line file is valid (gRPC port will be 0).
    Returns ``(0, 0)`` on any error.
    """
    dir_path = _data_dir()
    if not dir_path:
        return 0, 0

    port_file = Path(dir_path) / _PORT_FILE_NAME
    try:
        text = port_file.read_text(encoding="utf-8")
    except OSError:
        return 0, 0

    lines = text.strip().splitlines()
    if not lines:
        return 0, 0

    # Check PID staleness (line 3, if present)
    if len(lines) >= 3:
        pid = _parse_pid(lines[2])
        if pid is not None and not _is_process_alive(pid):
            return 0, 0

    rest_port = _parse_port(lines[0])
    grpc_port = _parse_port(lines[1]) if len(lines) >= 2 else 0
    return rest_port, grpc_port


def _parse_pid(s: str) -> int | None:
    """Parse a PID string, returning ``None`` if absent or invalid."""
    try:
        n = int(s.strip())
    except (ValueError, TypeError):
        return None
    if n > 0:
        return n
    return None


def _is_process_alive(pid: int) -> bool:
    """Return ``True`` if a process with *pid* is currently running."""
    try:
        os.kill(pid, 0)
    except ProcessLookupError:
        return False
    except PermissionError:
        # Process exists but we lack permission to signal it — still alive.
        return True
    except OSError:
        return False
    return True


def _parse_port(s: str) -> int:
    """Parse a port string, returning 0 on failure."""
    try:
        n = int(s.strip())
    except (ValueError, TypeError):
        return 0
    if 1 <= n <= 65535:
        return n
    return 0


def _data_dir() -> str:
    """Return the platform-specific data directory for the antd SDK daemon, or ``""``.

    The ``sdk`` subdirectory keeps antd's port file separate from the
    ant-node daemon, which writes to the same ``ant`` umbrella dir.
    """
    if sys.platform == "win32":
        appdata = os.environ.get("APPDATA", "")
        if not appdata:
            return ""
        return os.path.join(appdata, _DATA_DIR_NAME, _SDK_SUBDIR_NAME)

    if sys.platform == "darwin":
        home = os.environ.get("HOME", "")
        if not home:
            return ""
        return os.path.join(home, "Library", "Application Support", _DATA_DIR_NAME, _SDK_SUBDIR_NAME)

    # Linux and others
    xdg = os.environ.get("XDG_DATA_HOME", "")
    if xdg:
        return os.path.join(xdg, _DATA_DIR_NAME, _SDK_SUBDIR_NAME)
    home = os.environ.get("HOME", "")
    if not home:
        return ""
    return os.path.join(home, ".local", "share", _DATA_DIR_NAME, _SDK_SUBDIR_NAME)

File: ant-sdk/antd-ruby/lib/antd/discover.rb
# frozen_string_literal: true

require "rbconfig"

module Antd
  # Auto-discovers the antd daemon by reading the +daemon.port+ file that antd
  # writes on startup.
  #
  # The file contains up to three lines: REST port (line 1), gRPC port (line 2),
  # and optionally the daemon PID (line 3).
  #
  # If a PID is present and the process is no longer alive, the port file is
  # considered stale and discovery returns empty.
  #
  # Port file location is platform-specific:
  #   - Windows: %APPDATA%\ant\sdk\daemon.port
  #   - macOS:   ~/Library/Application Support/ant/sdk/daemon.port
  #   - Linux:   $XDG_DATA_HOME/ant/sdk/daemon.port or ~/.local/share/ant/sdk/daemon.port
  #
  # The +sdk+ subdirectory keeps antd's port file separate from the ant-node
  # daemon, which writes to the same +ant+ umbrella dir.
  module Discover
    PORT_FILE_NAME  = "daemon.port"
    DATA_DIR_NAME   = "ant"
    SDK_SUBDIR_NAME = "sdk"

    # Reads the daemon.port file and returns the REST base URL
    # (e.g. "http://127.0.0.1:8082").
    #
    # @return [String] the URL, or "" if the port file is not found
    def self.daemon_url
      rest, _ = read_port_file
      return "" if rest == 0

      "http://127.0.0.1:#{rest}"
    end

    # Reads the daemon.port file and returns the gRPC target
    # (e.g. "127.0.0.1:50051").
    #
    # @return [String] the target, or "" if the port file is not found
    def self.grpc_target
      _, grpc = read_port_file
      return "" if grpc == 0

      "127.0.0.1:#{grpc}"
    end

    # @api private
    def self.read_port_file
      dir = data_dir
      return [0, 0] if dir.empty?

      path = File.join(dir, PORT_FILE_NAME)
      return [0, 0] unless File.exist?(path)

      lines = File.read(path).strip.split("\n")
      rest_port = parse_port(lines[0])
      grpc_port = parse_port(lines[1])
      pid = parse_pid(lines[2])

      return [0, 0] if pid > 0 && !process_alive?(pid)

      [rest_port, grpc_port]
    rescue StandardError
      [0, 0]
    end

    # @api private
    def self.parse_port(str)
      return 0 if str.nil?

      s = str.strip
      return 0 unless s.match?(/\A\d+\z/)

      n = s.to_i
      (n > 0 && n <= 65535) ? n : 0
    end

    # @api private
    def self.data_dir
      host_os = RbConfig::CONFIG["host_os"]

      case host_os
      when /mswin|mingw|cygwin/
        appdata = ENV["APPDATA"]
        return "" if appdata.nil? || appdata.empty?

        File.join(appdata, DATA_DIR_NAME, SDK_SUBDIR_NAME)
      when /darwin/
        home = ENV["HOME"]
        return "" if home.nil? || home.empty?

        File.join(home, "Library", "Application Support", DATA_DIR_NAME, SDK_SUBDIR_NAME)
      else
        xdg = ENV["XDG_DATA_HOME"]
        if xdg && !xdg.empty?
          File.join(xdg, DATA_DIR_NAME, SDK_SUBDIR_NAME)
        else
          home = ENV["HOME"]
          return "" if home.nil? || home.empty?

          File.join(home, ".local", "share", DATA_DIR_NAME, SDK_SUBDIR_NAME)
        end
      end
    end

    # @api private
    def self.parse_pid(str)
      return 0 if str.nil?

      s = str.strip
      return 0 unless s.match?(/\A\d+\z/)

      n = s.to_i
      n > 0 ? n : 0
    end

    # @api private
    def self.process_alive?(pid)
      Process.kill(0, pid)
      true
    rescue Errno::ESRCH
      false
    rescue Errno::EPERM
      # Process exists but we lack permission to signal it — still alive.
      true
    end

    private_class_method :read_port_file, :parse_port, :parse_pid, :process_alive?, :data_dir
  end
end
